### DISCUSSION TITLE
Using an Azure DevOps Pull Request Template to remind your team about the Definition of Done

### DISCUSSION THREAD

**Discussion by giscus** [timdeschryver/timdeschryver.dev#274] [Announcements]
# Using an Azure DevOps Pull Request Template to remind your team about the Definition of Done

Making the team aware of the Definition of Done each time a Pull Request is created.

https://timdeschryver.dev/blog/using-an-azure-devops-pull-request-template-to-remind-your-team-about-the-definition-of-done

**Comment by markmcnulty** [comment]
This is great, although is there an additional repo setting to be able to use PR templates?

I have done the above but not seeing it in my PR description.

**Reply by timdeschryver** [reply]
This can't be enabled AFAIK, it should just work.
Make sure you have a `.azuredevops` folder in the root of your repository and that it a file named `pull_request_template.md`.

https://learn.microsoft.com/en-us/azure/devops/repos/git/pull-request-templates?view=azure-devops&WT.mc_id=DT-MVP-5004452

**Comment by wizofaus** [comment]
Can I change the template to include who the reviewers are?  Basically we want all PRs for each template to have the same set of reviewers.

**Reply by timdeschryver** [reply]
Yes, that's possible @wizofaus . You just need the user's id to be able to assign them.
For example, we extended the script to randomly assign a reviewer with the following script

```ps1
function AddReviewer {
    param(
        [string]$reviewerId,
        [string]$required
    )
    
    $url = "$($env:SYSTEM_TEAMFOUNDATIONCOLLECTIONURI)$env:SYSTEM_TEAMPROJECTID/_apis/git/repositories/$($env:REPOSITORY)/pullRequests/$($env:PR_ID)/reviewers/$($reviewerId)?api-version=7.0"
    $body = @{
        id = $reviewerId
        isRequired = $required
    } | ConvertTo-Json

    
    $pipeline = Invoke-RestMethod -Uri $url -Method 'Put' -Body $body -ContentType 'application/json' -Headers @{
        Authorization = "Bearer $env:SYSTEM_ACCESSTOKEN"
    }

    Write-Host "Pipeline = $($pipeline | ConvertTo-Json -Depth 100)"
}

function AssignRandomTeamMember {
    param(
        [string]$project
    )

    $teamUrl = "$($env:SYSTEM_TEAMFOUNDATIONCOLLECTIONURI)$env:SYSTEM_TEAMPROJECTID/_apis/projects/$($project)/teams/$($team)/members?api-version=7.0"
    $members = Invoke-RestMethod -Uri $teamUrl -Method 'GET' -Body $body -ContentType 'application/json' -Headers @{
        Authorization = "Bearer $env:SYSTEM_ACCESSTOKEN"
    }
    $reviewer = $members.value | Get-Random -Count 1 | Select-Object -ExpandProperty identity
    AddReviewer($reviewer.id, "false")
}

```

**Comment by wizofaus** [comment]
Awesome, thanks.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by Ungerfall** [comment]
Can the title be configured in the template?

**Reply by timdeschryver** [reply]
Can you elaborate why and what the title should be?
I don't think this is possible using the template, but I can be wrong, I never tried it.

One way to achieve this is by using the Azure DevOps API, similar to https://timdeschryver.dev/blog/invoke-the-azure-devops-api-from-within-your-azure-devops-pipeline.

**Reply by Ungerfall** [reply]
I would like to put a placeholder in the title "List related work items #12345".
This way the work items will be automatically linked to a pull request.

**Reply by timdeschryver** [reply]
Fyi, if you include the work item id in the commit message, then it will automatically be linked.

**Comment by coolraka** [comment]
How can we make a mandatory any one of the field so that while creating a PR user has to enter or checked that field otherwise it will throw an message and not giving you permission to create PR.

Thanks!!
